fix: issue #537

dbrennand opened the pull request:
Fixes #537

artis3n commented on tasks/centos/install.yml:
```diff
@@ -5,8 +5,21 @@
+- name: CentOS | Enable Tailscale Repo
+  become: true
+  when: tailscale_repo_state.stat.exists
+  ansible.builtin.command: dnf config-manager --enable tailscale-{{ release_stability }}
```
What if the disabled repo was a different `release_stability`?

dbrennand replied:
Good point! I just tested and `dnf config-manger` accepts a glob for the repo id:

`dnf config-manager --set-enabled tailscale*`

dbrennand replied:
Resolved in https://github.com/artis3n/ansible-role-tailscale/pull/538/commits/01b867124037c2527aae9635d02f98b46929805b

artis3n commented:
If you are familiar with Molecule, mind adding a scenario for the reproduction you describe in https://github.com/artis3n/ansible-role-tailscale/issues/537? That'll ensure this is captured in the CI test suite moving forward. If you're not familiar, no worries I can add that afterwards.

Also FYI that I am in the process of migrating this standarole role into a collection (https://github.com/artis3n/ansible-role-tailscale/discussions/529). The first release was just published this morning - https://github.com/artis3n/ansible-collection-tailscale.

We should make sure to add the changes from the final version of this PR over there as well.

dbrennand commented:
> If you are familiar with Molecule, mind adding a scenario for the reproduction you describe in #537? That'll ensure this is captured in the CI test suite moving forward. If you're not familiar, no worries I can add that afterwards.
> 
> Also FYI that I am in the process of migrating this standarole role into a collection (#529). The first release was just published this morning - https://github.com/artis3n/ansible-collection-tailscale.
> 
> We should make sure to add the changes from the final version of this PR over there as well.

Hi @artis3n - I've created a Molecule scenario for this in https://github.com/artis3n/ansible-role-tailscale/pull/538/commits/1cffc0fa9dbedcab16c8445620882c8e96193bac - Please could you take a look, see if it looks OK to you, I do see there is some idempotency test issues but some of the other scenarios I see don't run this sequence:

```
CRITICAL Idempotence test failed because of the following tasks:
*  => artis3n.tailscale : Uninstall | De-register Tailscale Node
*  => artis3n.tailscale : CentOS | Remove Tailscale
*  => artis3n.tailscale : Uninstall | Remove Tailscale Daemon State and Logs
*  => artis3n.tailscale : CentOS | Enable Tailscale Repo
*  => artis3n.tailscale : CentOS | Install Tailscale
*  => artis3n.tailscale : Install | Enable Service
*  => artis3n.tailscale : Install | Bring Tailscale Up
```

Apart from this, the behaviour is now as expected:

```
# Initial install
TASK [artis3n.tailscale : Install | CentOS and related families] ***************
included: /Users/dab/repos/forks/ansible-role-tailscale/tasks/centos/install.yml for instance

TASK [artis3n.tailscale : CentOS | DNF Dependencies] ***************************
changed: [instance]

TASK [artis3n.tailscale : CentOS | Check if Tailscale Repo is already present] ***
ok: [instance]

TASK [artis3n.tailscale : CentOS | Enable Tailscale Repo] **********************
skipping: [instance]

TASK [artis3n.tailscale : CentOS | Add Tailscale Repo] *************************
changed: [instance]

TASK [artis3n.tailscale : CentOS | Install Tailscale] **************************
changed: [instance]

# Reinstall after uninstall
TASK [Reinstall Tailscale using artis3n.tailscale] *****************************

TASK [artis3n.tailscale : Install | CentOS and related families] ***************
included: /Users/dab/repos/forks/ansible-role-tailscale/tasks/centos/install.yml for instance

TASK [artis3n.tailscale : CentOS | DNF Dependencies] ***************************
ok: [instance]

TASK [artis3n.tailscale : CentOS | Check if Tailscale Repo is already present] ***
ok: [instance]

TASK [artis3n.tailscale : CentOS | Enable Tailscale Repo] **********************
changed: [instance]

TASK [artis3n.tailscale : CentOS | Add Tailscale Repo] *************************
skipping: [instance]

TASK [artis3n.tailscale : CentOS | Install Tailscale] **************************
changed: [instance] 
```

artis3n commented on molecule/537/converge.yml:
```diff
@@ -0,0 +1,18 @@
+    - name: Init tailscale credentials variables
+      ansible.builtin.include_tasks: ../default/init_tailscale_vars.yml
+
+    - name: Remove Tailscale using artis3n.tailscale
+      ansible.builtin.include_role:
+        name: artis3n.tailscale
+      vars:
+        state: absent
```
There won't be any tailscale artifacts present. You should do an install, than an uninstall, and a reinstall.

Somewhat similar to this converge, although less complex than this one: https://github.com/artis3n/ansible-role-tailscale/blob/main/molecule/idempotent-up/converge.yml

dbrennand replied:
Hi @artis3n - I'm not sure I follow here, the scenario is doing an install, uninstall and then reinstall.

In the prepare.yml:

```yaml
    - name: Install latest Tailscale version using artis3n.tailscale
      ansible.builtin.include_role:
        name: artis3n.tailscale
      vars:
        state: latest
```

Then in the converge.yml

```yaml
    - name: Remove Tailscale using artis3n.tailscale
      ansible.builtin.include_role:
        name: artis3n.tailscale
      vars:
        state: absent

    - name: Reinstall Tailscale using artis3n.tailscale
      ansible.builtin.include_role:
        name: artis3n.tailscale
      vars:
        state: latest
```

artis3n replied:
Ahh, I see. I missed that was happening inside the prepare stage

dbrennand replied:
Are you ok with this? Or do you want it all in converge?

artis3n commented on molecule/537/molecule.yml:
```diff
@@ -0,0 +1,46 @@
+  name: docker
+platforms:
+  - name: instance
+    image: ${MOLECULE_DISTRO:-geerlingguy/docker-rockylinux9-ansible:latest}
```
Add a comment specifying why this default doesn't mirror the other scenarios using ubuntu 24.04 (because we're specifically testing a centos-like environment fix)

dbrennand replied:
Resolved in [392ae37](https://github.com/artis3n/ansible-role-tailscale/pull/538/commits/392ae371649f0a049e69054106ee5019d98536a6)

artis3n commented on molecule/537/molecule.yml:
```diff
@@ -0,0 +1,46 @@
+verifier:
+  name: ansible
+scenario:
+  name: "537"
```
Please update the directory name and the scenario name to "reinstall"

Right now centos-like distros are the only one with a reinstall bug, as you identified, but will use this scenario to encapsulate any reinstallation issues and ensure they are successful.

dbrennand replied:
Resolved in [392ae37](https://github.com/artis3n/ansible-role-tailscale/pull/538/commits/392ae371649f0a049e69054106ee5019d98536a6)

artis3n commented on tasks/centos/install.yml:
```diff
@@ -5,8 +5,21 @@
+    path: /etc/yum.repos.d/tailscale.repo
+  register: tailscale_repo_state
+
+- name: CentOS | Enable Tailscale Repo
+  become: true
+  when: tailscale_repo_state.stat.exists
+  ansible.builtin.command: "dnf config-manager --set-enabled tailscale*"
+  register: add_tailscale_repo
```
Thinking of how best we want to handle this idempotency check. `changed_when: true` doesn't make sense, as the yum repo might already be enabled so `--set-enabled` wouldn't constitute a change on the system.

Can we add a task? 
- Check if tailscale repo is already present
- Check if tailscale repo is already enabled
- Enable if disabled (this file will only run if `state` is `present` or `latest`)

![image](https://github.com/user-attachments/assets/ff263d38-b445-4132-bbb9-319db2bb2339)

dbrennand replied:
I need to work on this a bit more, I pushed some commits which have broken things currently.

dbrennand replied:
I'm unable to finish testing due to:

```
TASK [artis3n.tailscale : Warning: Role migrated into Collection] **************
ERROR! invalid meta action requested: end_role
```

dbrennand replied:
Fixed.

artis3n commented:
Oh, and also add this new scenario to CI by adding a job similar to this one, with your scenario name:

https://github.com/artis3n/ansible-role-tailscale/blob/main/.github/workflows/pull_request_target.yml#L116

artis3n commented:
You should have an option that says to allow edit by maintainers (https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) - mind checking that so I can push a few changes? I realized I forgot to add a couple of requests (e.g. make command) and while testing locally I applied some changes to the molecule scenario as well.

dbrennand commented:
Applie the requested changes @artis3n 🙂

artis3n reviewed (approved):
Thank you for your patience with this PR!

dbrennand commented:
@artis3n - Did you port these changes into the new collection?

artis3n commented:
@dbrennand I meant to - but I don't see this molecule test so I think I forgot 😬 I will do so tonight unless you get a PR up first
